chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -205,6 +210,8 @@ fn get_state_witness_block_range(
         }
         let (prev_shard_id, prev_shard_index) =
             epoch_manager.get_prev_shard_id(prev_hash, position.shard_id)?;
+        let prev_shard_layout =
+            epoch_manager.get_shard_layout(&epoch_manager.get_epoch_id(prev_hash)?)?;
```

Comment:
I'm not convinced that those are matching, the get_prev_shard_id logic is non-trivial. Perhaps you can refactor it to have the `get_prev_shard_id` return all `prev_shard_id`, `prev_shard_index` and `prev_shard_layout`?

Reply:
I'm not happy about extending return result even further, but... works for now.